drivers/adc/adc_cc32xx.c

```diff
@@ -0,0 +1,291 @@
+
+	data->repeat_buffer = data->buffer;
+	uint8_t ch = data->ch;
+
+	MAP_ADCChannelEnable(config->base, s_channel[ch]);
+	for (int i = 0; i < 5; i++) {
+		while (!MAP_ADCFIFOLvlGet(config->base, s_channel[ch])) {
+		}
+		MAP_ADCFIFORead(config->base, s_channel[ch]);
+	}
+	MAP_ADCIntClear(config->base, s_channel[ch], ISR_MASK);
+	MAP_ADCIntEnable(config->base, s_channel[ch], ISR_MASK);
+}
+
+static void adc_context_update_buffer_pointer(struct adc_context *ctx,
+					      bool repeat_sampling)
+{
+	struct adc_cc32xx_data *data =
+		CONTAINER_OF(ctx, struct adc_cc32xx_data, ctx);
+
+	if (repeat_sampling) {
+		data->buffer = data->repeat_buffer;
+	}
+}
+
+static int adc_cc32xx_init(const struct device *dev)
+{
+	struct adc_cc32xx_data *data = dev->data;
+	const struct adc_cc32xx_cfg *config = dev->config;
+
+	data->dev = dev;
+
+	LOG_DBG("Initializing....");
+
+	for (int i = 0; i < ARRAY_SIZE(s_channel); ++i) {
+		MAP_ADCIntDisable(config->base, s_channel[i], ISR_MASK);
+		MAP_ADCChannelDisable(config->base, s_channel[i]);
+		MAP_ADCDMADisable(config->base, s_channel[i]);
+		MAP_ADCIntClear(config->base, s_channel[i], ISR_MASK);
+	}
+	MAP_ADCEnable(config->base);
+	config->irq_cfg_func();
+
+	adc_context_unlock_unconditionally(&data->ctx);
+	return 0;
+}
+
+static int adc_cc32xx_channel_setup(const struct device *dev,
+				    const struct adc_channel_cfg *channel_cfg)
+{
+	struct adc_cc32xx_data *data = dev->data;
+	const struct adc_cc32xx_cfg *config = dev->config;
+	uint8_t ch = channel_cfg->channel_id;
+
+	if (ch >= ARRAY_SIZE(s_chPin)) {
+		LOG_ERR("Channel is not supported");
+		return -EINVAL;
+	}
+
+	if (channel_cfg->acquisition_time != ADC_ACQ_TIME_DEFAULT) {
+		LOG_ERR("Acquisition time is not valid");
+		return -EINVAL;
+	}
+
+	if (channel_cfg->differential) {
+		LOG_ERR("Differential channels are not supported");
+		return -EINVAL;
+	}
+
+	if (channel_cfg->gain != ADC_GAIN_1) {
+		LOG_ERR("Gain is not valid");
+		return -EINVAL;
+	}
+
+	if (channel_cfg->reference != ADC_REF_INTERNAL) {
+		LOG_ERR("Reference is not valid");
+		return -EINVAL;
+	}
+
+	data->ch = ch;
+
+	LOG_DBG("Setup %d", ch);
+
+	MAP_ADCChannelDisable(config->base, s_channel[ch]);
+	MAP_ADCIntDisable(config->base, s_channel[ch], ISR_MASK);
+	MAP_PinDirModeSet(s_chPin[ch], PIN_DIR_MODE_IN);
+	MAP_PinTypeADC(s_chPin[ch], PIN_MODE_255);
+
+	return 0;
+}
+
+static int cc32xx_read(const struct device *dev,
+		       const struct adc_sequence *sequence,
+		       bool asynchronous,
+		       struct k_poll_signal *sig)
+{
+	struct adc_cc32xx_data *data = dev->data;
+	uint32_t channels = sequence->channels;
+	uint32_t index = find_lsb_set(channels) - 1;
+	int rv;
+
+	if (sequence->resolution != 12) {
+		LOG_ERR("Resolution is not valid");
+		return -EINVAL;
+	}
+
+	if (channels > BIT(index)) {
+		LOG_ERR("Only single channel supported");
+		return -ENOTSUP;
+	}
+
+	data->buffer = sequence->buffer;
+	data->size = sequence->buffer_size / sizeof(*data->buffer);
```

Comment:
It should be checked here if the requested number of samples will fit in the provided buffer, otherwise users might get surprised. See for instance how it is done in [adc_nrfx_adc.c](https://github.com/zephyrproject-rtos/zephyr/blob/c1e1273df2a23329314a51023ac5f76069ecb5c4/drivers/adc/adc_nrfx_adc.c#L120-L137).

Reply:
Done